burn-import/src/onnx/state.rs

```diff
@@ -0,0 +1,246 @@
+        // Create a map to serialize the stateful nodes
+        let mut map = serializer.serialize_map(Some(stateful_nodes.len()))?;
+
+        // Serialize each stateful node into a map entry
+        for node in stateful_nodes.into_iter() {
+            match node.node_type {
+                NodeType::Conv2d => {
+                    let (name, record_item) = conv2d_state::<RS>(node);
+                    map.serialize_entry(&name, &record_item)?;
+                }
+                NodeType::Linear => {
+                    let (name, record_item) = linear_state::<RS>(node);
+                    map.serialize_entry(&name, &record_item)?;
+                }
+                NodeType::BatchNormalization => match first_input_dim(node).unwrap() {
+                    2 => {
+                        let (name, record_item) = batch_norm_state::<0, RS>(node);
+                        map.serialize_entry(&name, &record_item)?;
+                    }
+                    3 => {
+                        let (name, record_item) = batch_norm_state::<1, RS>(node);
+                        map.serialize_entry(&name, &record_item)?;
+                    }
+                    4 => {
+                        let (name, record_item) = batch_norm_state::<2, RS>(node);
+                        map.serialize_entry(&name, &record_item)?;
+                    }
+                    5 => {
+                        let (name, record_item) = batch_norm_state::<3, RS>(node);
+                        map.serialize_entry(&name, &record_item)?;
+                    }
+                    dim => todo!("BatchNorm for dim = {dim} is not implemented yet"),
+                },
+
+                _ => todo!(
+                    "Serialize state for node type {:?} is not implemented yet",
+                    node.node_type
+                ),
+            }
+        }
+        map.end()
+    }
+}
+
+/// Dummy implementation of Deserialize for ModelState
+impl<'de, RS: RecordSettings> Deserialize<'de> for ModelState<RS> {
+    fn deserialize<D>(_deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        unimplemented!() // No intention to deserialize a ModelState
+    }
+}
+
+/// Convert itermediate representation of tensor into a burn tensor
+///
+/// TODO: implemenate for all tensor element types
+impl<const D: usize> TryFrom<&ir::Tensor> for Tensor<B, D> {
+    type Error = ();
+
+    fn try_from(value: &ir::Tensor) -> Result<Self, Self::Error> {
+        let shape: [usize; D] = value.shape.clone().try_into().unwrap();
+        let TensorData::Float32s(floats) = value.data.clone().unwrap() else {
+            todo!("Tensor data must be float32s");
+        };
+        let tensor: Tensor<B, D> = Tensor::from_data(floats.as_slice()).reshape(shape);
+        Ok(tensor)
+    }
+}
+
+/// Convert a Conv2d node into a Name, Conv2dRecordItem pair
+pub fn conv2d_state<RS: RecordSettings>(node: &Node) -> (String, Conv2dRecordItem<B, RS>) {
```

Comment:
Because we are serializing here so we have to have Item, which uses Serialize trait.

Reply:
I don't think we should serialize here, just returning the `Record` should be fine. Fixing this will probably resolve other comments that I have about seriazation. 

In other words, the place where we serialize and do type conversion should change. Work as long as possible with `Record` and avoid using `Item<S>`, idealy it should only be used by the recorder when `record` is called.